File: src/python/prompt_formatter.py
from typing import List, Dict, Any, Optional
from schemas import Message, File


messages_prompt = "The following message interactions have occured"
messages_tag = "messages"
message_total_char_limit = 5000 # how many character are included in a message chunk accross all messages

sleep_tag = "sleep_consolidation"

file_tag = "file"
file_part_tag = "file_part"
file_char_limit = 20000 # how many character to include in file part 

def format_messages(messages: List[Message]) -> List[str]:
    """
    Format messages like the following: 

    user: hi my name is sarah
    assistant: Hello Sarah! I'm Sam. smiles warmly There's something special about first meetings, don't you think? Like opening a book to its first page, full of possibilities. I'd love to get to know you better - what brings you here today?
    assistant: Tool call returned Sent message successfully.

    """
    message_history = "\n".join([
        f"{msg.role}: {msg.content}" for msg in messages
    ])

    return [{"role": "user", "content": f"<{messages_tag}>{messages_prompt}:\n{message_history}</{messages_tag}>"}]

def format_sleep_prompt(
    blocks: List[Any],
    archival_passages: Optional[List[str]] = None,
) -> List[Dict[str, str]]:
    """
    Format a sleep/consolidation prompt containing the current memory state.

    The prompt asks the agent to review its own memory blocks and consolidate them:
    resolve contradictions, merge redundancies, prune stale info, and reorganize.
    """
    # Format current block state
    block_lines = []
    for block in blocks:
        label = getattr(block, "label", None)
        description = getattr(block, "description", "")
        value = getattr(block, "value", "")
        if label is None:
            continue
        block_lines.append(
            f'<block label="{label}" description="{description}">\n{value}\n</block>'
        )
    blocks_section = "\n\n".join(block_lines)

    # Optional archival section
    archival_section = ""
    if archival_passages:
        passages_text = "\n".join(f"- {p}" for p in archival_passages)
        archival_section = (
            f"\n\nThe following are recent passages from archival (long-term) memory. "
            f"Consider whether any of this information should be promoted into your "
            f"active memory blocks, or whether it reveals patterns worth capturing:\n"
            f"{passages_text}"
        )

    prompt = (
        f"<{sleep_tag}>\n"
        f"You are entering a sleep/consolidation phase. No new external information "
        f"is being provided. Instead, review your current memory state and improve it.\n\n"
        f"Your tasks:\n"
        f"1. Identify and resolve any contradictions between memory blocks\n"
        f"2. Merge redundant information that appears across multiple blocks\n"
        f"3. Remove or condense stale or outdated information\n"
        f"4. Strengthen connections between related facts across blocks\n"
        f"5. Reorganize information within blocks for clarity and coherence\n"
        f"6. Note any gaps in your knowledge that future conversations should address\n\n"
        f"Current memory blocks:\n{blocks_section}"
        f"{archival_section}\n\n"
        f"Update your memory blocks to reflect a consolidated, coherent understanding. "
        f"Do not fabricate new information — only reorganize and refine what you already know.\n"
        f"</{sleep_tag}>"
    )

    return [{"role": "user", "content": prompt}]


def format_files(files: List[File]) -> List[Dict[str, str]]:
    """
    Format files into multiple separate messages, ensuring each message stays under file_char_limit.
    Each file part becomes its own message.
    """
    all_messages = []
    
    for file in files:
        try:
            # Read file content from disk
            with open(file.file_path, 'r', encoding='utf-8') as f:
                file_content = f.read()
        except Exception as e:
            # If we can't read the file, send an error message
            error_msg = f"<{file_tag} label=\"{file.label}\" description=\"{file.description}\">[Error reading file: {str(e)}]</{file_tag}>"
            all_messages.append({"role": "user", "content": error_msg})
            continue
            
        # chunk up file content into parts 
        file_content_chunks = [file_content[i:i + file_char_limit] for i in range(0, len(file_content), file_char_limit)]
        
        # Create a separate message for each file part
        for i, chunk in enumerate(file_content_chunks):
            part_number = i + 1
            total_parts = len(file_content_chunks)
            
            # Create the message content for this specific part
            file_part_content = f"<{file_part_tag} part={part_number}/{total_parts}>{chunk}</{file_part_tag}>"
            file_message = f"<{file_tag} label=\"{file.label}\" description=\"{file.description}\">{file_part_content}</{file_tag}>"
            
            all_messages.append({"role": "user", "content": file_message})
    
    return all_messages
    

    

File: src/typescript/src/prompt-formatter.ts
import * as fs from 'fs';
import { Message, File, MessageCreate } from './schemas';

const MESSAGES_PROMPT = "The following message interactions have occured";
const MESSAGES_TAG = "messages";
const MESSAGE_TOTAL_CHAR_LIMIT = 5000;

const FILE_TAG = "file";
const FILE_PART_TAG = "file_part";
const FILE_CHAR_LIMIT = 20000;

const SLEEP_TAG = "sleep_consolidation";

export function formatMessages(messages: MessageCreate[]): Array<{ role: string; content: string }> {
  const messageHistory = messages
    .map(msg => `${msg.role}: ${msg.content}`)
    .join('\n');

  return [{
    role: "user",
    content: `<${MESSAGES_TAG}>${MESSAGES_PROMPT}:\n${messageHistory}</${MESSAGES_TAG}>`
  }];
}

export function formatSleepPrompt(
  blocks: any[],
  archivalPassages?: string[] | null,
): Array<{ role: string; content: string }> {
  const blockLines = blocks
    .filter((b: any) => b.label)
    .map((b: any) => {
      const label = b.label;
      const description = b.description || '';
      const value = b.value || '';
      return `<block label="${label}" description="${description}">\n${value}\n</block>`;
    });
  const blocksSection = blockLines.join('\n\n');

  let archivalSection = '';
  if (archivalPassages && archivalPassages.length > 0) {
    const passagesText = archivalPassages.map(p => `- ${p}`).join('\n');
    archivalSection =
      `\n\nThe following are recent passages from archival (long-term) memory. ` +
      `Consider whether any of this information should be promoted into your ` +
      `active memory blocks, or whether it reveals patterns worth capturing:\n` +
      passagesText;
  }

  const prompt =
    `<${SLEEP_TAG}>\n` +
    `You are entering a sleep/consolidation phase. No new external information ` +
    `is being provided. Instead, review your current memory state and improve it.\n\n` +
    `Your tasks:\n` +
    `1. Identify and resolve any contradictions between memory blocks\n` +
    `2. Merge redundant information that appears across multiple blocks\n` +
    `3. Remove or condense stale or outdated information\n` +
    `4. Strengthen connections between related facts across blocks\n` +
    `5. Reorganize information within blocks for clarity and coherence\n` +
    `6. Note any gaps in your knowledge that future conversations should address\n\n` +
    `Current memory blocks:\n${blocksSection}` +
    `${archivalSection}\n\n` +
    `Update your memory blocks to reflect a consolidated, coherent understanding. ` +
    `Do not fabricate new information — only reorganize and refine what you already know.\n` +
    `</${SLEEP_TAG}>`;

  return [{ role: 'user', content: prompt }];
}

export function formatFiles(files: File[]): Array<{ role: string; content: string }> {
  const allMessages: Array<{ role: string; content: string }> = [];

  for (const file of files) {
    try {
      const fileContent = fs.readFileSync(file.file_path, 'utf-8');
      
      const fileContentChunks = [];
      for (let i = 0; i < fileContent.length; i += FILE_CHAR_LIMIT) {
        fileContentChunks.push(fileContent.slice(i, i + FILE_CHAR_LIMIT));
      }

      console.log(`Formatted file ${file.label} into ${fileContentChunks.length} separate messages`);

      for (let i = 0; i < fileContentChunks.length; i++) {
        const chunk = fileContentChunks[i];
        const partNumber = i + 1;
        const totalParts = fileContentChunks.length;

        const filePartContent = `<${FILE_PART_TAG} part=${partNumber}/${totalParts}>${chunk}</${FILE_PART_TAG}>`;
        const fileMessage = `<${FILE_TAG} label="${file.label}" description="${file.description}">${filePartContent}</${FILE_TAG}>`;

        allMessages.push({ role: "user", content: fileMessage });
      }
    } catch (error) {
      const errorMsg = `<${FILE_TAG} label="${file.label}" description="${file.description}">[Error reading file: ${error}]</${FILE_TAG}>`;
      allMessages.push({ role: "user", content: errorMsg });
    }
  }

  return allMessages;
}